### PR TITLE
[ALLUXIO-2743] Support specifying Mount Options for each Mount Point (part4)

### PR DESCRIPTION
**Pull request by apc999**
https://alluxio.atlassian.net/browse/ALLUXIO-2743

This PR updates each UFS adaptor to take advantage from the ufs conf in Map to create the UFS client libraries.

A follow up PR to https://github.com/Alluxio/alluxio/pull/5217/files, you can review them in parallel.

### CODE REVIEW

File: core/common/src/main/java/alluxio/util/UnderFileSystemUtils.java [hunk 1]
```diff
@@ -158,5 +161,30 @@ public static boolean isSwift(UnderFileSystem ufs) {
     return "swift".equals(ufs.getUnderFSType());
   }
 
+  /**
+   * @param key property key
+   * @param ufsConf ufs configuration
+   * @return the value of the given key in ufs conf, or global conf if the key found in ufs conf
```
Comment (jsimsa) [inline core/common/src/main/java/alluxio/util/UnderFileSystemUtils.java]: if the is *not* found

File: underfs/gcs/src/main/java/alluxio/underfs/gcs/GCSUnderFileSystem.java [hunk 2]
```diff
@@ -78,27 +79,30 @@
    * Constructs a new instance of {@link GCSUnderFileSystem}.
    *
    * @param uri the {@link AlluxioURI} for this UFS
+   * @param ufsConf the configuration for this ufs
```
Comment (jsimsa) [inline underfs/gcs/src/main/java/alluxio/underfs/gcs/GCSUnderFileSystem.java]: nit: use consistent capitalization of UFS (the line above uses "UFS", you use "ufs")

File: underfs/glusterfs/src/main/java/alluxio/underfs/glusterfs/GlusterFSUnderFileSystem.java [hunk 3]
```diff
@@ -27,45 +31,68 @@
...
   }
+
+  /**
+   * Factory method to construct a new Gluster FS {@link UnderFileSystem}.
+   *
+   * @param uri the {@link AlluxioURI} for this UFS
+   * @param conf the configuration for Hadoop or GlusterFS
+   * @return a new Gluster FS {@link UnderFileSystem}
+   */
+  public static GlusterFSUnderFileSystem createInstance(AlluxioURI uri, Map<String, String> conf) {
```
Comment (jsimsa) [inline underfs/glusterfs/src/main/java/alluxio/underfs/glusterfs/GlusterFSUnderFileSystem.java]: nit: call this method simply `create`

Comment (jsimsa) [inline underfs/glusterfs/src/main/java/alluxio/underfs/glusterfs/GlusterFSUnderFileSystem.java]: possibly wrapping it in a nested `Factory` class

File: underfs/hdfs/src/main/java/alluxio/underfs/hdfs/HdfsUnderFileSystem.java [hunk 4]
```diff
@@ -61,34 +62,38 @@
...
-    if (conf != null && conf instanceof org.apache.hadoop.conf.Configuration) {
-      hadoopConf = (org.apache.hadoop.conf.Configuration) conf;
-    } else {
-      hadoopConf = new org.apache.hadoop.conf.Configuration();
-    }
-    prepareConfiguration(ufsPrefix, hadoopConf);
-    hadoopConf.addResource(
-        new Path(hadoopConf.get(PropertyKey.UNDERFS_HDFS_CONFIGURATION.toString())));
-    HdfsUnderFileSystemUtils.addS3Credentials(hadoopConf);
+  public static HdfsUnderFileSystem createInstance(AlluxioURI ufsUri, Map<String, String> conf) {
```
Comment (jsimsa) [inline underfs/hdfs/src/main/java/alluxio/underfs/hdfs/HdfsUnderFileSystem.java]: similar comment as for GCS

File: underfs/hdfs/src/main/java/alluxio/underfs/hdfs/HdfsUnderFileSystem.java [hunk 5]
```diff
@@ -61,34 +62,38 @@
...
+    Configuration hadoopConf = createConfiguration(conf);
+    return new HdfsUnderFileSystem(ufsUri, conf, hadoopConf);
+  }
 
-    Path path = new Path(ufsPrefix);
+  /**
+   * Constructs a new HDFS {@link UnderFileSystem}.
+   *
+   * @param ufsUri the {@link AlluxioURI} for this UFS
+   * @param ufsConf the configuration for ufs
```
Comment (jsimsa) [inline underfs/hdfs/src/main/java/alluxio/underfs/hdfs/HdfsUnderFileSystem.java]: use `this UFS` consistently

File: underfs/glusterfs/src/main/java/alluxio/underfs/glusterfs/GlusterFSUnderFileSystem.java [hunk 6]
```diff
@@ -27,45 +31,68 @@
...
+    Configuration hadoopConf = prepareConfiguration(uri.toString(), conf);
+    return new GlusterFSUnderFileSystem(uri, conf, hadoopConf);
+  }
+
+  /**
+   * Constructs a new Gluster FS {@link UnderFileSystem}.
+   *
+   * @param ufsUri the {@link AlluxioURI} for this UFS
+   * @param ufsConf the configuration for ufs
```
Comment (jsimsa) [inline underfs/glusterfs/src/main/java/alluxio/underfs/glusterfs/GlusterFSUnderFileSystem.java]: use `this UFS` consistently

File: core/common/src/main/java/alluxio/util/UnderFileSystemUtils.java (hunk 1 above)
Comment (apc999) [inline core/common/src/main/java/alluxio/util/UnderFileSystemUtils.java]: done

File: underfs/gcs/src/main/java/alluxio/underfs/gcs/GCSUnderFileSystem.java (hunk 2 above)
Comment (apc999) [inline underfs/gcs/src/main/java/alluxio/underfs/gcs/GCSUnderFileSystem.java]: done

File: underfs/glusterfs/src/main/java/alluxio/underfs/glusterfs/GlusterFSUnderFileSystem.java (hunk 3 above)
Comment (apc999) [inline underfs/glusterfs/src/main/java/alluxio/underfs/glusterfs/GlusterFSUnderFileSystem.java]: it seems the convention in all UFS adaptors is to call the factory method `createInstance`. I am open to either `createInstance` or `create` (I personally prefer `create`), but I think we should change the name consistently. 

Regarding `Factory class`, we have already `ClusterFSUnderFileSystemFactory` class. I would avoid having two factory classes as it is confusing for us in the future which one to use.

Comment (apc999) [inline underfs/glusterfs/src/main/java/alluxio/underfs/glusterfs/GlusterFSUnderFileSystem.java]: Let me know if you want to change `createInstance` across different ufs adaptors, if yes, I could change them in this PR.

File: underfs/glusterfs/src/main/java/alluxio/underfs/glusterfs/GlusterFSUnderFileSystem.java (hunk 6 above)
Comment (apc999) [inline underfs/glusterfs/src/main/java/alluxio/underfs/glusterfs/GlusterFSUnderFileSystem.java]: done

File: underfs/hdfs/src/main/java/alluxio/underfs/hdfs/HdfsUnderFileSystem.java (hunk 4 above)
Comment (apc999) [inline underfs/hdfs/src/main/java/alluxio/underfs/hdfs/HdfsUnderFileSystem.java]: replied in GCS

File: underfs/glusterfs/src/main/java/alluxio/underfs/glusterfs/GlusterFSUnderFileSystem.java (hunk 3 above)
Comment (jsimsa) [inline underfs/glusterfs/src/main/java/alluxio/underfs/glusterfs/GlusterFSUnderFileSystem.java]: no need, let's keep it consistent and not slow you down

File: underfs/hdfs/src/main/java/alluxio/underfs/hdfs/HdfsUnderFileSystem.java (hunk 5 above)
Comment (apc999) [inline underfs/hdfs/src/main/java/alluxio/underfs/hdfs/HdfsUnderFileSystem.java]: done

**Comment by apc999** [comment]
@jsimsa comments addressed. PTAL

File: core/common/src/main/java/alluxio/underfs/UnderFileSystem.java [hunk 7]
```diff
@@ -68,7 +68,7 @@ private Cache() {}
        * @param ufsConf the ufs configuration
        * @return the UFS instance
        */
-      UnderFileSystem get(String path, Object ufsConf) {
+      UnderFileSystem get(String path, Map<String, String> ufsConf) {
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/underfs/UnderFileSystem.java]: Does it make sense to have a wrapper for this in case we want to add other functionality?

File: core/common/src/main/java/alluxio/util/UnderFileSystemUtils.java [hunk 8]
```diff
@@ -158,5 +161,34 @@ public static boolean isSwift(UnderFileSystem ufs) {
...
+
+  /**
+   * @param key property key
+   * @param ufsConf configuration for the UFS
+   * @return true if the key is contained in the given UFS configuration or global configuration
+   */
+  public static boolean containsKey(PropertyKey key, Map<String, String> ufsConf) {
+    return (ufsConf != null && ufsConf.containsKey(key.toString())) || Configuration
+        .containsKey(key);
+  }
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/util/UnderFileSystemUtils.java]: These two methods seem out of place, it does not seem intuitive to include this logic in `UnderFileSystemUtils`. This seems more like trying to merge configurations or lookup a value given a configuration chain which is unrelated to the under file system.

Comment (apc999) [inline core/common/src/main/java/alluxio/util/UnderFileSystemUtils.java]: Correct. Agree this is a workaround. Also that is why I was asking when we could have the instanced configuration, as I think the cleaner way is to extend the instanced configuration class as UFS configuration which performs a chained configuration lookup based on the passed-in conf and global conf. 

I could implement a class like this, the concern is it is likely to be changed again after we finalize the configuration story. So I created two util methods as workaround. are you strongly against the util methods?  @calvinjia

File: core/common/src/main/java/alluxio/underfs/UnderFileSystem.java (hunk 7 above)
Comment (apc999) [inline core/common/src/main/java/alluxio/underfs/UnderFileSystem.java]: do you mean create a class sth like `UfsConf`, which is simply a wrapper of `Map<String, String>`?

Comment (calvinjia) [inline core/common/src/main/java/alluxio/underfs/UnderFileSystem.java]: Yeah, passing a raw map here seems hard to extend in the future, if we need some other qualifiers from the configuration.

File: core/common/src/main/java/alluxio/util/UnderFileSystemUtils.java (hunk 8 above)
Comment (calvinjia) [inline core/common/src/main/java/alluxio/util/UnderFileSystemUtils.java]: I'm okay with the util methods for now, I think they are not in the correct utils class, since their functionality is not restricted to the `UnderFileSystem`. If we have a wrapper for the map, I think that would be better. It would also be good to make it more intuitive that this method checks both the conf argument and falls back to the Alluxio conf. Just looking at the method name & signature my expectations are quite different.

Comment (apc999) [inline core/common/src/main/java/alluxio/util/UnderFileSystemUtils.java]: @calvinjia SGTM. How about making a separate PR on this? the reason is currently both part3 and part4 heavily depend on the ufsConf type. If I made a change in this PR, the other part3 PR needs to be updated too, introducing a lot overhead. I would suggest to have a separate PR after part3 and part4 merged. 

For now, I will leave a TODO.

File: core/common/src/main/java/alluxio/underfs/UnderFileSystem.java (hunk 7 above)
Comment (apc999) [inline core/common/src/main/java/alluxio/underfs/UnderFileSystem.java]: as I commented in below, I agree with the change and suggest to have a separate PR on this.

File: core/common/src/main/java/alluxio/util/UnderFileSystemUtils.java (hunk 8 above)
Comment (apc999) [inline core/common/src/main/java/alluxio/util/UnderFileSystemUtils.java]: Alternatively, I could wait for part3 merged and then come back to this PR. It is up to you.

Comment (calvinjia) [inline core/common/src/main/java/alluxio/util/UnderFileSystemUtils.java]: Let's leave the change for a separate PR.
